# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [20]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [21]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [22]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    source_sentences = source_text.lower().split('\n')
    target_sentences = target_text.lower().split('\n')
    target_sentences = [sentence + ' <EOS>' for sentence in target_sentences]
    
    source_int = [[source_vocab_to_int[word] for word in sentence.split()] 
                  for sentence in source_sentences]
    target_int = [[target_vocab_to_int[word] for word in sentence.split()]
                  for sentence in target_sentences]
    return source_int, target_int

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)


Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [23]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [24]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [25]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.2.1
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [37]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    inputs = tf.placeholder(tf.int32, [None, None], "input")
    targets = tf.placeholder(tf.int32, [None, None], "targets")
    learn_rate = tf.placeholder(tf.float32, None, "learning_rate")
    keep_prob = tf.placeholder(tf.float32, None, "keep_prob")
    target_seq_len = tf.placeholder(tf.int32, (None,), "target_sequence_length")
    max_target_len = tf.reduce_max(target_seq_len, name='max_target_len')
    source_sequence_len = tf.placeholder(tf.int32, (None,), "source_sequence_length")
    
    return inputs, targets, learn_rate, keep_prob, target_seq_len, max_target_len, source_sequence_len


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Operation'>):
<tf.Operation 'assert_rank_2/Assert/Assert' type=Assert>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
['File "C:\\ProgramData\\Anaconda3\\envs\\tf_1_2\\lib\\runpy.py", line 193, in _run_module_as_main\n    "__main__", mod_spec)', 'File "C:\\ProgramData\\Anaconda3\\envs\\tf_1_2\\lib\\runpy.py", line 85, in _run_code\n    exec(code, run_globals)', 'File "C:\\ProgramData\\Anaconda3\\envs\\tf_1_2\\lib\\site-packages\\ipykernel\\__main__.py", line 3, in <module>\n    app.launch_new_instance()', 'File "C:\\ProgramData\\Anaconda3\\envs\\tf_1_2\\lib\\site-packages\\traitlets\\config\\application.py", line 658, in launch_instance\n    app.start()', 'File "C:\\ProgramData\\Anaconda3\\envs\\tf_1_2\\lib\\site-packages\\ipykernel\\kernelapp.py", line 477, in start\n    ioloop.IOLoop.instance().start()', 'File "C:\\Prog

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [38]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # TODO: Implement Function
    x = tf.strided_slice(target_data, [0,0], [batch_size, -1], [1, 1])
    y = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), x], 1)
            
    return y

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [39]:
def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # TODO: Implement Function
    enc_embed_input = tf.contrib.layers.embed_sequence(rnn_inputs, source_vocab_size, encoding_embedding_size)
    
    def make_cell(rnn_size):
        enc_cell = tf.contrib.rnn.LSTMCell(rnn_size,
                                           initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        return enc_cell
    enc_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
    
    enc_output, enc_state = tf.nn.dynamic_rnn(enc_cell, enc_embed_input, sequence_length=source_sequence_length, dtype=tf.float32)
    
    return enc_output, enc_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [40]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    # TODO: Implement Function


    # Helper for the training process. Used by BasicDecoder to read inputs.
    training_helper = tf.contrib.seq2seq.TrainingHelper(inputs=dec_embed_input, sequence_length=target_sequence_length,
                                                        time_major=False)
    
    # Basic decoder
    training_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, training_helper, encoder_state, output_layer) 

    # Perform dynamic decoding using the decoder
    training_decoder_output = tf.contrib.seq2seq.dynamic_decode(training_decoder, impute_finished=True,
                                                                   maximum_iterations=max_summary_length)

    return training_decoder_output[0]


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [41]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    # TODO: Implement Function
    
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')

    # Helper for the inference process.
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(dec_embeddings, start_tokens, end_of_sequence_id)

    # Basic decoder
    inference_decoder = tf.contrib.seq2seq.BasicDecoder(dec_cell, inference_helper, encoder_state, output_layer)

    # Perform dynamic decoding using the decoder
    inference_decoder_output = tf.contrib.seq2seq.dynamic_decode(inference_decoder, impute_finished=True,
                                                        maximum_iterations=max_target_sequence_length)

    return inference_decoder_output[0]

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [42]:
def decoding_layer(dec_input, encoder_state, target_sequence_length, max_target_sequence_length,
                   rnn_size, num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    # 1. Decoder Embedding
    
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)

    # 2. Construct the decoder cell
    def make_cell(rnn_size):
        dec_cell = tf.contrib.rnn.LSTMCell(rnn_size,
                                           initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        return dec_cell

    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])
     
    # 3. Dense layer to translate the decoder's output at each time 
    # step into a choice from the target vocabulary
    output_layer = Dense(target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))


    with tf.variable_scope("decode"):
        training_decoder_output = decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, 
                                                max_target_sequence_length, output_layer, keep_prob)
    
    with tf.variable_scope("decode", reuse=True):
        infer_decoder_output = decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, target_vocab_to_int['<GO>'], 
                                                    target_vocab_to_int['<EOS>'], max_target_sequence_length, target_vocab_size, 
                                                    output_layer, batch_size, keep_prob)
        
    
    return training_decoder_output, infer_decoder_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [43]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sequence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    # Pass the input data through the encoder. We'll ignore the encoder output, but use the state
    _, enc_state = encoding_layer(input_data, rnn_size, num_layers, keep_prob, source_sequence_length, source_vocab_size, 
                                  enc_embedding_size)
    
    # Prepare the target sequences we'll feed to the decoder in training mode
    dec_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    # Pass encoder state and decoder inputs to the decoders
    training_decoder_output, inference_decoder_output = decoding_layer(dec_input, enc_state, target_sequence_length, 
                                                                       max_target_sequence_length, rnn_size, num_layers, 
                                                                       target_vocab_to_int, target_vocab_size, batch_size, 
                                                                       keep_prob, dec_embedding_size)
    
    return training_decoder_output, inference_decoder_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [44]:
# Number of Epochs
epochs = 20
# Batch Size
batch_size = 256
# RNN Size
rnn_size = 256
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 256
decoding_embedding_size = 256
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.5
display_step = 10

### Build the Graph
Build the graph using the neural network you implemented.

In [45]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [46]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [47]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   10/1077 - Train Accuracy: 0.2582, Validation Accuracy: 0.3583, Loss: 3.4539
Epoch   0 Batch   20/1077 - Train Accuracy: 0.3434, Validation Accuracy: 0.4002, Loss: 2.7898
Epoch   0 Batch   30/1077 - Train Accuracy: 0.3793, Validation Accuracy: 0.4325, Loss: 2.5786
Epoch   0 Batch   40/1077 - Train Accuracy: 0.4301, Validation Accuracy: 0.4723, Loss: 2.3802
Epoch   0 Batch   50/1077 - Train Accuracy: 0.4461, Validation Accuracy: 0.4961, Loss: 2.2649
Epoch   0 Batch   60/1077 - Train Accuracy: 0.5048, Validation Accuracy: 0.5295, Loss: 1.9625
Epoch   0 Batch   70/1077 - Train Accuracy: 0.4979, Validation Accuracy: 0.5497, Loss: 1.8726
Epoch   0 Batch   80/1077 - Train Accuracy: 0.5316, Validation Accuracy: 0.5820, Loss: 1.6398
Epoch   0 Batch   90/1077 - Train Accuracy: 0.5199, Validation Accuracy: 0.5732, Loss: 1.5395
Epoch   0 Batch  100/1077 - Train Accuracy: 0.5246, Validation Accuracy: 0.5543, Loss: 1.3141
Epoch   0 Batch  110/1077 - Train Accuracy: 0.5672, Validati

Epoch   0 Batch  890/1077 - Train Accuracy: 0.9498, Validation Accuracy: 0.9229, Loss: 0.0509
Epoch   0 Batch  900/1077 - Train Accuracy: 0.9574, Validation Accuracy: 0.9332, Loss: 0.0606
Epoch   0 Batch  910/1077 - Train Accuracy: 0.9501, Validation Accuracy: 0.9251, Loss: 0.0472
Epoch   0 Batch  920/1077 - Train Accuracy: 0.9090, Validation Accuracy: 0.9219, Loss: 0.0468
Epoch   0 Batch  930/1077 - Train Accuracy: 0.9719, Validation Accuracy: 0.9034, Loss: 0.0360
Epoch   0 Batch  940/1077 - Train Accuracy: 0.9527, Validation Accuracy: 0.9237, Loss: 0.0413
Epoch   0 Batch  950/1077 - Train Accuracy: 0.9382, Validation Accuracy: 0.9169, Loss: 0.0431
Epoch   0 Batch  960/1077 - Train Accuracy: 0.9431, Validation Accuracy: 0.9222, Loss: 0.0400
Epoch   0 Batch  970/1077 - Train Accuracy: 0.9570, Validation Accuracy: 0.9464, Loss: 0.0516
Epoch   0 Batch  980/1077 - Train Accuracy: 0.9297, Validation Accuracy: 0.9290, Loss: 0.0470
Epoch   0 Batch  990/1077 - Train Accuracy: 0.9507, Validati

Epoch   1 Batch  700/1077 - Train Accuracy: 0.9680, Validation Accuracy: 0.9400, Loss: 0.0279
Epoch   1 Batch  710/1077 - Train Accuracy: 0.9539, Validation Accuracy: 0.9354, Loss: 0.0268
Epoch   1 Batch  720/1077 - Train Accuracy: 0.9544, Validation Accuracy: 0.9364, Loss: 0.0398
Epoch   1 Batch  730/1077 - Train Accuracy: 0.9574, Validation Accuracy: 0.9311, Loss: 0.0358
Epoch   1 Batch  740/1077 - Train Accuracy: 0.9480, Validation Accuracy: 0.9425, Loss: 0.0264
Epoch   1 Batch  750/1077 - Train Accuracy: 0.9645, Validation Accuracy: 0.9414, Loss: 0.0254
Epoch   1 Batch  760/1077 - Train Accuracy: 0.9488, Validation Accuracy: 0.9311, Loss: 0.0416
Epoch   1 Batch  770/1077 - Train Accuracy: 0.9777, Validation Accuracy: 0.9467, Loss: 0.0312
Epoch   1 Batch  780/1077 - Train Accuracy: 0.9508, Validation Accuracy: 0.9371, Loss: 0.0494
Epoch   1 Batch  790/1077 - Train Accuracy: 0.9145, Validation Accuracy: 0.9435, Loss: 0.0435
Epoch   1 Batch  800/1077 - Train Accuracy: 0.9520, Validati

Epoch   2 Batch  510/1077 - Train Accuracy: 0.9418, Validation Accuracy: 0.9570, Loss: 0.0395
Epoch   2 Batch  520/1077 - Train Accuracy: 0.9732, Validation Accuracy: 0.9570, Loss: 0.0243
Epoch   2 Batch  530/1077 - Train Accuracy: 0.9578, Validation Accuracy: 0.9379, Loss: 0.0309
Epoch   2 Batch  540/1077 - Train Accuracy: 0.9504, Validation Accuracy: 0.9222, Loss: 0.0251
Epoch   2 Batch  550/1077 - Train Accuracy: 0.9500, Validation Accuracy: 0.9205, Loss: 0.0272
Epoch   2 Batch  560/1077 - Train Accuracy: 0.9398, Validation Accuracy: 0.9403, Loss: 0.0394
Epoch   2 Batch  570/1077 - Train Accuracy: 0.9420, Validation Accuracy: 0.9329, Loss: 0.0453
Epoch   2 Batch  580/1077 - Train Accuracy: 0.9565, Validation Accuracy: 0.9379, Loss: 0.0232
Epoch   2 Batch  590/1077 - Train Accuracy: 0.9613, Validation Accuracy: 0.9244, Loss: 0.0395
Epoch   2 Batch  600/1077 - Train Accuracy: 0.9702, Validation Accuracy: 0.9442, Loss: 0.0371
Epoch   2 Batch  610/1077 - Train Accuracy: 0.9638, Validati

Epoch   3 Batch  320/1077 - Train Accuracy: 0.9602, Validation Accuracy: 0.9656, Loss: 0.0389
Epoch   3 Batch  330/1077 - Train Accuracy: 0.9477, Validation Accuracy: 0.9098, Loss: 0.0332
Epoch   3 Batch  340/1077 - Train Accuracy: 0.9790, Validation Accuracy: 0.9268, Loss: 0.0284
Epoch   3 Batch  350/1077 - Train Accuracy: 0.9684, Validation Accuracy: 0.9237, Loss: 0.0386
Epoch   3 Batch  360/1077 - Train Accuracy: 0.9594, Validation Accuracy: 0.9450, Loss: 0.0265
Epoch   3 Batch  370/1077 - Train Accuracy: 0.9613, Validation Accuracy: 0.9496, Loss: 0.0276
Epoch   3 Batch  380/1077 - Train Accuracy: 0.9738, Validation Accuracy: 0.9666, Loss: 0.0228
Epoch   3 Batch  390/1077 - Train Accuracy: 0.9727, Validation Accuracy: 0.9474, Loss: 0.0356
Epoch   3 Batch  400/1077 - Train Accuracy: 0.9543, Validation Accuracy: 0.9343, Loss: 0.0406
Epoch   3 Batch  410/1077 - Train Accuracy: 0.9527, Validation Accuracy: 0.9513, Loss: 0.0361
Epoch   3 Batch  420/1077 - Train Accuracy: 0.9812, Validati

Epoch   4 Batch  130/1077 - Train Accuracy: 0.9751, Validation Accuracy: 0.9268, Loss: 0.0220
Epoch   4 Batch  140/1077 - Train Accuracy: 0.9807, Validation Accuracy: 0.9265, Loss: 0.0258
Epoch   4 Batch  150/1077 - Train Accuracy: 0.9587, Validation Accuracy: 0.9247, Loss: 0.0296
Epoch   4 Batch  160/1077 - Train Accuracy: 0.9887, Validation Accuracy: 0.9357, Loss: 0.0189
Epoch   4 Batch  170/1077 - Train Accuracy: 0.9566, Validation Accuracy: 0.9418, Loss: 0.0369
Epoch   4 Batch  180/1077 - Train Accuracy: 0.9691, Validation Accuracy: 0.9347, Loss: 0.0260
Epoch   4 Batch  190/1077 - Train Accuracy: 0.9609, Validation Accuracy: 0.9268, Loss: 0.0333
Epoch   4 Batch  200/1077 - Train Accuracy: 0.9793, Validation Accuracy: 0.9343, Loss: 0.0288
Epoch   4 Batch  210/1077 - Train Accuracy: 0.9468, Validation Accuracy: 0.9350, Loss: 0.0369
Epoch   4 Batch  220/1077 - Train Accuracy: 0.9646, Validation Accuracy: 0.9517, Loss: 0.0365
Epoch   4 Batch  230/1077 - Train Accuracy: 0.9740, Validati

Epoch   4 Batch 1010/1077 - Train Accuracy: 0.9637, Validation Accuracy: 0.9393, Loss: 0.0277
Epoch   4 Batch 1020/1077 - Train Accuracy: 0.9578, Validation Accuracy: 0.9535, Loss: 0.0279
Epoch   4 Batch 1030/1077 - Train Accuracy: 0.9855, Validation Accuracy: 0.9521, Loss: 0.0349
Epoch   4 Batch 1040/1077 - Train Accuracy: 0.9716, Validation Accuracy: 0.9279, Loss: 0.0228
Epoch   4 Batch 1050/1077 - Train Accuracy: 0.9664, Validation Accuracy: 0.9386, Loss: 0.0243
Epoch   4 Batch 1060/1077 - Train Accuracy: 0.9770, Validation Accuracy: 0.9425, Loss: 0.0218
Epoch   4 Batch 1070/1077 - Train Accuracy: 0.9750, Validation Accuracy: 0.9425, Loss: 0.0262
Epoch   5 Batch   10/1077 - Train Accuracy: 0.9790, Validation Accuracy: 0.9531, Loss: 0.0273
Epoch   5 Batch   20/1077 - Train Accuracy: 0.9691, Validation Accuracy: 0.9517, Loss: 0.0288
Epoch   5 Batch   30/1077 - Train Accuracy: 0.9637, Validation Accuracy: 0.9435, Loss: 0.0256
Epoch   5 Batch   40/1077 - Train Accuracy: 0.9773, Validati

Epoch   5 Batch  820/1077 - Train Accuracy: 0.9648, Validation Accuracy: 0.9535, Loss: 0.0330
Epoch   5 Batch  830/1077 - Train Accuracy: 0.9230, Validation Accuracy: 0.9229, Loss: 0.0336
Epoch   5 Batch  840/1077 - Train Accuracy: 0.9629, Validation Accuracy: 0.9478, Loss: 0.0273
Epoch   5 Batch  850/1077 - Train Accuracy: 0.9494, Validation Accuracy: 0.9560, Loss: 0.0518
Epoch   5 Batch  860/1077 - Train Accuracy: 0.9632, Validation Accuracy: 0.9609, Loss: 0.0331
Epoch   5 Batch  870/1077 - Train Accuracy: 0.9774, Validation Accuracy: 0.9577, Loss: 0.0228
Epoch   5 Batch  880/1077 - Train Accuracy: 0.9785, Validation Accuracy: 0.9634, Loss: 0.0343
Epoch   5 Batch  890/1077 - Train Accuracy: 0.9706, Validation Accuracy: 0.9354, Loss: 0.0275
Epoch   5 Batch  900/1077 - Train Accuracy: 0.9598, Validation Accuracy: 0.9506, Loss: 0.0312
Epoch   5 Batch  910/1077 - Train Accuracy: 0.9520, Validation Accuracy: 0.9283, Loss: 0.0339
Epoch   5 Batch  920/1077 - Train Accuracy: 0.9648, Validati

Epoch   6 Batch  630/1077 - Train Accuracy: 0.9711, Validation Accuracy: 0.9496, Loss: 0.0354
Epoch   6 Batch  640/1077 - Train Accuracy: 0.9594, Validation Accuracy: 0.9524, Loss: 0.0316
Epoch   6 Batch  650/1077 - Train Accuracy: 0.9723, Validation Accuracy: 0.9403, Loss: 0.0296
Epoch   6 Batch  660/1077 - Train Accuracy: 0.9727, Validation Accuracy: 0.9418, Loss: 0.0231
Epoch   6 Batch  670/1077 - Train Accuracy: 0.9709, Validation Accuracy: 0.9542, Loss: 0.0280
Epoch   6 Batch  680/1077 - Train Accuracy: 0.9483, Validation Accuracy: 0.9347, Loss: 0.0333
Epoch   6 Batch  690/1077 - Train Accuracy: 0.9656, Validation Accuracy: 0.9403, Loss: 0.0325
Epoch   6 Batch  700/1077 - Train Accuracy: 0.9672, Validation Accuracy: 0.9386, Loss: 0.0227
Epoch   6 Batch  710/1077 - Train Accuracy: 0.9543, Validation Accuracy: 0.9471, Loss: 0.0274
Epoch   6 Batch  720/1077 - Train Accuracy: 0.9663, Validation Accuracy: 0.9606, Loss: 0.0337
Epoch   6 Batch  730/1077 - Train Accuracy: 0.9746, Validati

Epoch   7 Batch  440/1077 - Train Accuracy: 0.9535, Validation Accuracy: 0.9418, Loss: 0.0297
Epoch   7 Batch  450/1077 - Train Accuracy: 0.9602, Validation Accuracy: 0.9421, Loss: 0.0333
Epoch   7 Batch  460/1077 - Train Accuracy: 0.9633, Validation Accuracy: 0.9510, Loss: 0.0359
Epoch   7 Batch  470/1077 - Train Accuracy: 0.9646, Validation Accuracy: 0.9517, Loss: 0.0264
Epoch   7 Batch  480/1077 - Train Accuracy: 0.9609, Validation Accuracy: 0.9471, Loss: 0.0277
Epoch   7 Batch  490/1077 - Train Accuracy: 0.9613, Validation Accuracy: 0.9332, Loss: 0.0234
Epoch   7 Batch  500/1077 - Train Accuracy: 0.9867, Validation Accuracy: 0.9396, Loss: 0.0222
Epoch   7 Batch  510/1077 - Train Accuracy: 0.9641, Validation Accuracy: 0.9521, Loss: 0.0266
Epoch   7 Batch  520/1077 - Train Accuracy: 0.9855, Validation Accuracy: 0.9460, Loss: 0.0228
Epoch   7 Batch  530/1077 - Train Accuracy: 0.9316, Validation Accuracy: 0.9432, Loss: 0.0319
Epoch   7 Batch  540/1077 - Train Accuracy: 0.9563, Validati

Epoch   8 Batch  250/1077 - Train Accuracy: 0.9485, Validation Accuracy: 0.9513, Loss: 0.0276
Epoch   8 Batch  260/1077 - Train Accuracy: 0.9751, Validation Accuracy: 0.9513, Loss: 0.0230
Epoch   8 Batch  270/1077 - Train Accuracy: 0.9633, Validation Accuracy: 0.9467, Loss: 0.0266
Epoch   8 Batch  280/1077 - Train Accuracy: 0.9453, Validation Accuracy: 0.9492, Loss: 0.0317
Epoch   8 Batch  290/1077 - Train Accuracy: 0.9664, Validation Accuracy: 0.9329, Loss: 0.0386
Epoch   8 Batch  300/1077 - Train Accuracy: 0.9778, Validation Accuracy: 0.9435, Loss: 0.0180
Epoch   8 Batch  310/1077 - Train Accuracy: 0.9719, Validation Accuracy: 0.9290, Loss: 0.0284
Epoch   8 Batch  320/1077 - Train Accuracy: 0.9742, Validation Accuracy: 0.9414, Loss: 0.0323
Epoch   8 Batch  330/1077 - Train Accuracy: 0.9469, Validation Accuracy: 0.9361, Loss: 0.0313
Epoch   8 Batch  340/1077 - Train Accuracy: 0.9782, Validation Accuracy: 0.9251, Loss: 0.0275
Epoch   8 Batch  350/1077 - Train Accuracy: 0.9609, Validati

Epoch   9 Batch   60/1077 - Train Accuracy: 0.9669, Validation Accuracy: 0.9254, Loss: 0.0233
Epoch   9 Batch   70/1077 - Train Accuracy: 0.9408, Validation Accuracy: 0.9343, Loss: 0.0286
Epoch   9 Batch   80/1077 - Train Accuracy: 0.9676, Validation Accuracy: 0.9332, Loss: 0.0255
Epoch   9 Batch   90/1077 - Train Accuracy: 0.9684, Validation Accuracy: 0.9400, Loss: 0.0155
Epoch   9 Batch  100/1077 - Train Accuracy: 0.9766, Validation Accuracy: 0.9492, Loss: 0.0245
Epoch   9 Batch  110/1077 - Train Accuracy: 0.9555, Validation Accuracy: 0.9503, Loss: 0.0217
Epoch   9 Batch  120/1077 - Train Accuracy: 0.9648, Validation Accuracy: 0.9574, Loss: 0.0305
Epoch   9 Batch  130/1077 - Train Accuracy: 0.9688, Validation Accuracy: 0.9641, Loss: 0.0225
Epoch   9 Batch  140/1077 - Train Accuracy: 0.9770, Validation Accuracy: 0.9354, Loss: 0.0250
Epoch   9 Batch  150/1077 - Train Accuracy: 0.9859, Validation Accuracy: 0.9411, Loss: 0.0293
Epoch   9 Batch  160/1077 - Train Accuracy: 0.9828, Validati

Epoch   9 Batch  940/1077 - Train Accuracy: 0.9629, Validation Accuracy: 0.9521, Loss: 0.0260
Epoch   9 Batch  950/1077 - Train Accuracy: 0.9576, Validation Accuracy: 0.9347, Loss: 0.0254
Epoch   9 Batch  960/1077 - Train Accuracy: 0.9754, Validation Accuracy: 0.9400, Loss: 0.0254
Epoch   9 Batch  970/1077 - Train Accuracy: 0.9773, Validation Accuracy: 0.9446, Loss: 0.0320
Epoch   9 Batch  980/1077 - Train Accuracy: 0.9602, Validation Accuracy: 0.9311, Loss: 0.0258
Epoch   9 Batch  990/1077 - Train Accuracy: 0.9560, Validation Accuracy: 0.9276, Loss: 0.0415
Epoch   9 Batch 1000/1077 - Train Accuracy: 0.9624, Validation Accuracy: 0.9350, Loss: 0.0282
Epoch   9 Batch 1010/1077 - Train Accuracy: 0.9789, Validation Accuracy: 0.9592, Loss: 0.0229
Epoch   9 Batch 1020/1077 - Train Accuracy: 0.9766, Validation Accuracy: 0.9432, Loss: 0.0187
Epoch   9 Batch 1030/1077 - Train Accuracy: 0.9812, Validation Accuracy: 0.9446, Loss: 0.0268
Epoch   9 Batch 1040/1077 - Train Accuracy: 0.9383, Validati

Epoch  10 Batch  750/1077 - Train Accuracy: 0.9656, Validation Accuracy: 0.9478, Loss: 0.0253
Epoch  10 Batch  760/1077 - Train Accuracy: 0.9734, Validation Accuracy: 0.9464, Loss: 0.0352
Epoch  10 Batch  770/1077 - Train Accuracy: 0.9747, Validation Accuracy: 0.9474, Loss: 0.0165
Epoch  10 Batch  780/1077 - Train Accuracy: 0.9340, Validation Accuracy: 0.9471, Loss: 0.0409
Epoch  10 Batch  790/1077 - Train Accuracy: 0.9441, Validation Accuracy: 0.9503, Loss: 0.0423
Epoch  10 Batch  800/1077 - Train Accuracy: 0.9414, Validation Accuracy: 0.9656, Loss: 0.0275
Epoch  10 Batch  810/1077 - Train Accuracy: 0.9758, Validation Accuracy: 0.9379, Loss: 0.0193
Epoch  10 Batch  820/1077 - Train Accuracy: 0.9781, Validation Accuracy: 0.9478, Loss: 0.0177
Epoch  10 Batch  830/1077 - Train Accuracy: 0.9484, Validation Accuracy: 0.9418, Loss: 0.0320
Epoch  10 Batch  840/1077 - Train Accuracy: 0.9859, Validation Accuracy: 0.9517, Loss: 0.0201
Epoch  10 Batch  850/1077 - Train Accuracy: 0.9594, Validati

Epoch  11 Batch  560/1077 - Train Accuracy: 0.9680, Validation Accuracy: 0.9322, Loss: 0.0256
Epoch  11 Batch  570/1077 - Train Accuracy: 0.9424, Validation Accuracy: 0.9425, Loss: 0.0501
Epoch  11 Batch  580/1077 - Train Accuracy: 0.9650, Validation Accuracy: 0.9638, Loss: 0.0162
Epoch  11 Batch  590/1077 - Train Accuracy: 0.9589, Validation Accuracy: 0.9485, Loss: 0.0290
Epoch  11 Batch  600/1077 - Train Accuracy: 0.9568, Validation Accuracy: 0.9332, Loss: 0.0336
Epoch  11 Batch  610/1077 - Train Accuracy: 0.9704, Validation Accuracy: 0.9489, Loss: 0.0351
Epoch  11 Batch  620/1077 - Train Accuracy: 0.9527, Validation Accuracy: 0.9517, Loss: 0.0311
Epoch  11 Batch  630/1077 - Train Accuracy: 0.9844, Validation Accuracy: 0.9634, Loss: 0.0265
Epoch  11 Batch  640/1077 - Train Accuracy: 0.9847, Validation Accuracy: 0.9517, Loss: 0.0213
Epoch  11 Batch  650/1077 - Train Accuracy: 0.9703, Validation Accuracy: 0.9577, Loss: 0.0338
Epoch  11 Batch  660/1077 - Train Accuracy: 0.9801, Validati

Epoch  12 Batch  370/1077 - Train Accuracy: 0.9777, Validation Accuracy: 0.9563, Loss: 0.0274
Epoch  12 Batch  380/1077 - Train Accuracy: 0.9891, Validation Accuracy: 0.9464, Loss: 0.0191
Epoch  12 Batch  390/1077 - Train Accuracy: 0.9625, Validation Accuracy: 0.9300, Loss: 0.0378
Epoch  12 Batch  400/1077 - Train Accuracy: 0.9508, Validation Accuracy: 0.9428, Loss: 0.0404
Epoch  12 Batch  410/1077 - Train Accuracy: 0.9539, Validation Accuracy: 0.9624, Loss: 0.0394
Epoch  12 Batch  420/1077 - Train Accuracy: 0.9793, Validation Accuracy: 0.9659, Loss: 0.0201
Epoch  12 Batch  430/1077 - Train Accuracy: 0.9715, Validation Accuracy: 0.9609, Loss: 0.0201
Epoch  12 Batch  440/1077 - Train Accuracy: 0.9516, Validation Accuracy: 0.9400, Loss: 0.0333
Epoch  12 Batch  450/1077 - Train Accuracy: 0.9645, Validation Accuracy: 0.9439, Loss: 0.0345
Epoch  12 Batch  460/1077 - Train Accuracy: 0.9574, Validation Accuracy: 0.9478, Loss: 0.0320
Epoch  12 Batch  470/1077 - Train Accuracy: 0.9663, Validati

Epoch  13 Batch  180/1077 - Train Accuracy: 0.9570, Validation Accuracy: 0.9407, Loss: 0.0218
Epoch  13 Batch  190/1077 - Train Accuracy: 0.9797, Validation Accuracy: 0.9549, Loss: 0.0247
Epoch  13 Batch  200/1077 - Train Accuracy: 0.9691, Validation Accuracy: 0.9513, Loss: 0.0231
Epoch  13 Batch  210/1077 - Train Accuracy: 0.9598, Validation Accuracy: 0.9442, Loss: 0.0341
Epoch  13 Batch  220/1077 - Train Accuracy: 0.9618, Validation Accuracy: 0.9581, Loss: 0.0272
Epoch  13 Batch  230/1077 - Train Accuracy: 0.9602, Validation Accuracy: 0.9648, Loss: 0.0317
Epoch  13 Batch  240/1077 - Train Accuracy: 0.9742, Validation Accuracy: 0.9389, Loss: 0.0253
Epoch  13 Batch  250/1077 - Train Accuracy: 0.9648, Validation Accuracy: 0.9492, Loss: 0.0269
Epoch  13 Batch  260/1077 - Train Accuracy: 0.9769, Validation Accuracy: 0.9496, Loss: 0.0152
Epoch  13 Batch  270/1077 - Train Accuracy: 0.9469, Validation Accuracy: 0.9542, Loss: 0.0379
Epoch  13 Batch  280/1077 - Train Accuracy: 0.9355, Validati

Epoch  13 Batch 1060/1077 - Train Accuracy: 0.9477, Validation Accuracy: 0.9496, Loss: 0.0248
Epoch  13 Batch 1070/1077 - Train Accuracy: 0.9531, Validation Accuracy: 0.9446, Loss: 0.0211
Epoch  14 Batch   10/1077 - Train Accuracy: 0.9527, Validation Accuracy: 0.9482, Loss: 0.0314
Epoch  14 Batch   20/1077 - Train Accuracy: 0.9738, Validation Accuracy: 0.9606, Loss: 0.0225
Epoch  14 Batch   30/1077 - Train Accuracy: 0.9766, Validation Accuracy: 0.9648, Loss: 0.0211
Epoch  14 Batch   40/1077 - Train Accuracy: 0.9770, Validation Accuracy: 0.9734, Loss: 0.0247
Epoch  14 Batch   50/1077 - Train Accuracy: 0.9719, Validation Accuracy: 0.9702, Loss: 0.0236
Epoch  14 Batch   60/1077 - Train Accuracy: 0.9594, Validation Accuracy: 0.9368, Loss: 0.0177
Epoch  14 Batch   70/1077 - Train Accuracy: 0.9527, Validation Accuracy: 0.9585, Loss: 0.0266
Epoch  14 Batch   80/1077 - Train Accuracy: 0.9754, Validation Accuracy: 0.9656, Loss: 0.0246
Epoch  14 Batch   90/1077 - Train Accuracy: 0.9680, Validati

Epoch  14 Batch  870/1077 - Train Accuracy: 0.9749, Validation Accuracy: 0.9467, Loss: 0.0198
Epoch  14 Batch  880/1077 - Train Accuracy: 0.9797, Validation Accuracy: 0.9641, Loss: 0.0268
Epoch  14 Batch  890/1077 - Train Accuracy: 0.9587, Validation Accuracy: 0.9563, Loss: 0.0204
Epoch  14 Batch  900/1077 - Train Accuracy: 0.9691, Validation Accuracy: 0.9592, Loss: 0.0352
Epoch  14 Batch  910/1077 - Train Accuracy: 0.9695, Validation Accuracy: 0.9545, Loss: 0.0216
Epoch  14 Batch  920/1077 - Train Accuracy: 0.9789, Validation Accuracy: 0.9581, Loss: 0.0190
Epoch  14 Batch  930/1077 - Train Accuracy: 0.9625, Validation Accuracy: 0.9496, Loss: 0.0208
Epoch  14 Batch  940/1077 - Train Accuracy: 0.9543, Validation Accuracy: 0.9442, Loss: 0.0324
Epoch  14 Batch  950/1077 - Train Accuracy: 0.9542, Validation Accuracy: 0.9400, Loss: 0.0232
Epoch  14 Batch  960/1077 - Train Accuracy: 0.9747, Validation Accuracy: 0.9521, Loss: 0.0202
Epoch  14 Batch  970/1077 - Train Accuracy: 0.9637, Validati

Epoch  15 Batch  680/1077 - Train Accuracy: 0.9580, Validation Accuracy: 0.9400, Loss: 0.0320
Epoch  15 Batch  690/1077 - Train Accuracy: 0.9590, Validation Accuracy: 0.9570, Loss: 0.0260
Epoch  15 Batch  700/1077 - Train Accuracy: 0.9691, Validation Accuracy: 0.9549, Loss: 0.0264
Epoch  15 Batch  710/1077 - Train Accuracy: 0.9582, Validation Accuracy: 0.9567, Loss: 0.0266
Epoch  15 Batch  720/1077 - Train Accuracy: 0.9819, Validation Accuracy: 0.9418, Loss: 0.0284
Epoch  15 Batch  730/1077 - Train Accuracy: 0.9684, Validation Accuracy: 0.9450, Loss: 0.0353
Epoch  15 Batch  740/1077 - Train Accuracy: 0.9434, Validation Accuracy: 0.9293, Loss: 0.0310
Epoch  15 Batch  750/1077 - Train Accuracy: 0.9508, Validation Accuracy: 0.9332, Loss: 0.0214
Epoch  15 Batch  760/1077 - Train Accuracy: 0.9680, Validation Accuracy: 0.9283, Loss: 0.0411
Epoch  15 Batch  770/1077 - Train Accuracy: 0.9624, Validation Accuracy: 0.9528, Loss: 0.0230
Epoch  15 Batch  780/1077 - Train Accuracy: 0.9348, Validati

Epoch  16 Batch  490/1077 - Train Accuracy: 0.9539, Validation Accuracy: 0.9421, Loss: 0.0304
Epoch  16 Batch  500/1077 - Train Accuracy: 0.9602, Validation Accuracy: 0.9567, Loss: 0.0219
Epoch  16 Batch  510/1077 - Train Accuracy: 0.9352, Validation Accuracy: 0.9471, Loss: 0.0292
Epoch  16 Batch  520/1077 - Train Accuracy: 0.9821, Validation Accuracy: 0.9244, Loss: 0.0200
Epoch  16 Batch  530/1077 - Train Accuracy: 0.9531, Validation Accuracy: 0.9268, Loss: 0.0288
Epoch  16 Batch  540/1077 - Train Accuracy: 0.9520, Validation Accuracy: 0.9506, Loss: 0.0302
Epoch  16 Batch  550/1077 - Train Accuracy: 0.9457, Validation Accuracy: 0.9652, Loss: 0.0269
Epoch  16 Batch  560/1077 - Train Accuracy: 0.9633, Validation Accuracy: 0.9453, Loss: 0.0237
Epoch  16 Batch  570/1077 - Train Accuracy: 0.9498, Validation Accuracy: 0.9386, Loss: 0.0390
Epoch  16 Batch  580/1077 - Train Accuracy: 0.9635, Validation Accuracy: 0.9375, Loss: 0.0171
Epoch  16 Batch  590/1077 - Train Accuracy: 0.9634, Validati

KeyboardInterrupt: 

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [ ]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [ ]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [ ]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    return None


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

## Translate
This will translate `translate_sentence` from English to French.

In [ ]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.